Added base server

pcodes opened the pull request:
- Ported base server code from @Others to work for the router
- Added initial structure for a ComponentRequest

pcodes committed:
Massive initial commit, future ones will be better

- Ported base server code from @Others to work for the router
- Added initial structure for a ComponentRequest

Others commented on Cargo.toml:
```diff
@@ -0,0 +1,16 @@
+name = "v9_router"
+version = "0.1.0"
+authors = ["Pat Cody <patcody028@gmail.com>"]
+edition = "2018"
```
Add more information here -- see the v9_worker Cargo.toml

Others commented on src/error.rs:
```diff
@@ -0,0 +1,64 @@
+        }
+        Ok(())
+    }
+}
```
Rewrite this using the `write!` macro pls

Others commented on src/main.rs:
```diff
@@ -0,0 +1,37 @@
+#![warn(clippy::pedantic, clippy::needless_borrow)]
+// But I don't care about these ones for now (most applicable since the code isn't fleshed out)
+#![allow(clippy::module_name_repetitions)]
+#![allow(clippy::needless_pass_by_value)]
```
Update to this:
```
// I'd like the most pedantic warning level
#![warn(
    clippy::cargo,
    clippy::needless_borrow,
    clippy::pedantic,
    clippy::redundant_clone
)]
// But I don't care about these ones
#![allow(
    clippy::cast_precision_loss,     // There is no way to avoid this precision loss
    clippy::module_name_repetitions, // Sometimes clear naming calls for repetition
    clippy::multiple_crate_versions,  // There is no way to easily fix this without modifying our dependencies
    clippy::needless_pass_by_value // FIXME: Remove once the code is in a better state 
)]
```
Then rerun `cargo clippy`

Others commented on src/main.rs:
```diff
@@ -0,0 +1,37 @@
+#[macro_use]
+extern crate log;
+
+extern crate serde;
```
If you aren't using macros, you don't need this line

pcodes replied:
Line 9? I do need the `#[macro_use]` otherwise it doesn't compile

Others replied:
Line 20

Others replied:
Line 20

Others commented on src/main.rs:
```diff
@@ -0,0 +1,37 @@
+mod error;
+mod request_handler;
+mod router;
+mod server;
```
Should be
```
extern crate
...
mod
...
use
```

Others commented on src/main.rs:
```diff
@@ -0,0 +1,37 @@
+mod server;
+
+fn main() {
+    let is_development_mode = dbg!(env::args()).any(|arg| arg == "--development");
```
Remove `dbg`

Others commented on src/request_handler.rs:
```diff
@@ -0,0 +1,93 @@
+use crate::error::RouterError;
+use crate::router::ComponentRequest;
+
+//mod router;
```
Delete

Others commented on src/error.rs:
```diff
@@ -0,0 +1,64 @@
+    fn from(e: serde_json::Error) -> Self {
+        RouterError::InternalJsonHandling(e)
+    }
+}
```
Weird ordering... Put all the `From`s together

Others commented on src/error.rs:
```diff
@@ -0,0 +1,64 @@
+            .status(StatusCode::INTERNAL_SERVER_ERROR)
+            .body(Body::from(self.to_string()))
+            .unwrap()
+    }
```
This does not make sense for you. You need to return a unique error code for when your code is broken

Others commented on src/request_handler.rs:
```diff
@@ -0,0 +1,93 @@
+            error!("INTERNAL SERVER ERROR -- {:?}", resp);
+        } else {
+            debug!("{:?}", resp);
+        }
```
See comment above -- you should rethink your error codes

Others commented on src/request_handler.rs:
```diff
@@ -0,0 +1,93 @@
+        let repo = path_components[2].to_string();
+        let method = path_components[3].to_string();
+
+        let request = ComponentRequest::new(http_verb, query, body, user, repo, method);
+        let result_body = request.forward_component_request();
```
You probably want a separate object that is in charge of forwarding and has all the node info ect.

This does not block this PR tho

Others commented on src/error.rs:
```diff
@@ -0,0 +1,64 @@
+
+impl Display for RouterError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+        // Technically this isn't very DRY, but I felt like factoring it out hurt readability YMMV
```
Delete comment

Others commented on src/main.rs:
```diff
@@ -0,0 +1,37 @@
+fn main() {
+    let is_development_mode = dbg!(env::args()).any(|arg| arg == "--development");
+    if is_development_mode {
+        println!("Starting in development mode");
```
log, don't print directly

Others commented on src/main.rs:
```diff
@@ -0,0 +1,37 @@
+mod router;
+mod server;
+
+fn main() {
```
Actually set up your logger here

Others reviewed (changes requested):
Looks basically good, just a few small things :)

pcodes committed:
Updated Cargo.toml to include missing info

pcodes committed:
Replaced error printing with write! macro

pcodes committed:
Cleaned up main.rs

pcodes committed:
Minor code clean-up

pcodes committed:
Modified error codes

- Router will now return a 532 if there is an internal router error vs
a component error

pcodes committed:
Set up logger

pcodes commented:
Should be ready to merge @Others

pcodes committed:
My code is pretty now
